dbms/src/Flash/Mpp/ExchangeReceiver.cpp

```diff
@@ -342,21 +328,25 @@ class AsyncRequestHandler : public UnaryCallback<bool>
     void start()
     {
         stage = AsyncRequestStage::WAIT_MAKE_READER;
+
         // Use lock to ensure async reader is unreachable from grpc thread before this function returns
         std::unique_lock lock(mu);
-        rpc_context->makeAsyncReader(*request, reader, thisAsUnaryCallback());
+        rpc_context->makeAsyncReader(*request, reader, cq, thisAsUnaryCallback());
     }
 
-    void waitForRetryOrDone(String done_msg)
+    void retryOrDone(String done_msg)
     {
         if (retriable())
         {
-            expected_retry_ts = Clock::now() + std::chrono::seconds(1);
-            stage = AsyncRequestStage::WAIT_RETRY;
-            reader.reset();
+            // Let alarm put me into CompletionQueue after a while
+            // , so that we can try to connect again.
+            alarm.Set(cq, Clock::now() + std::chrono::seconds(1), this);
```

Comment:
After setting the alarm, this struct may be popped from completion queue in another thread. So there may be a data race between `++retry_times;` and another function call.